umbra-js/src/classes/Umbra.ts

```diff
@@ -235,6 +256,119 @@ export class Umbra {
+      const token = batch[i].token;
+      const amount = batch[i].amount;
+      if (tokenSum.get(token) == undefined) {
+        tokenSum.set(token, amount);
+        tokens.push(token);
+      } else {
+        tokenSum.set(token, tokenSum.get(token).add(amount));
+      }
+    }
+
+    // If applicable, check that sender has sufficient token balance. ETH balance is checked on send. The isEth
+    // method also serves to validate the token input
+    for (let j = 0; j < tokens.length; j++) {
+      const token = tokens[j];
+      if (!isEth(token)) {
+        const tokenContract = new Contract(token, ERC20_ABI, signer) as ERC20;
+        const tokenBalance = await tokenContract.balanceOf(await signer.getAddress());
+        if (tokenBalance.lt(tokenSum.get(token))) {
+          const providedAmount = BigNumber.from(tokenSum.get(token)).toString();
+          const details = `Has ${tokenBalance.toString()} tokens, tried to send ${providedAmount} tokens.`;
+          throw new Error(`Insufficient balance to complete transfer. ${details}`);
+        }
+      }
+    }
+
+    const txSigner = this.getConnectedSigner(signer); // signer input validated
+
+    // Get toll amount from contract
+    const toll = await this.umbraContract.toll();
+
+    // Parse provided overrides
+    const localOverrides = { ...overrides }; // avoid mutating the object passed in
+    const advanced = localOverrides?.advanced || false;
+    const supportPubKey = localOverrides?.supportPubKey || false;
+    const supportTxHash = localOverrides?.supportTxHash || false;
+    const lookupOverrides = { advanced, supportPubKey, supportTxHash };
+
+    delete localOverrides.advanced;
+    delete localOverrides.supportPubKey;
+    delete localOverrides.supportTxHash;
+
+    let sendData: SendData[] = [];
+    let valueAmount: BigNumber = BigNumber.from('0');
+    let stealthKeyPairs: KeyPair[] = [];
+
+    for (let i = 0; i < batch.length; i++) {
+      // Lookup recipient's public key
+      const { spendingPublicKey, viewingPublicKey } = await lookupRecipient(
+        batch[i].address,
+        this.provider,
+        lookupOverrides
+      );
+      if (!spendingPublicKey || !viewingPublicKey) {
+        throw new Error(`Could not retrieve public keys for recipient ID ${batch[i].address}`);
+      }
+
+      const token = batch[i].token;
+      const amount = batch[i].amount;
+
+      const spendingKeyPair = new KeyPair(spendingPublicKey);
+      const viewingKeyPair = new KeyPair(viewingPublicKey);
+
+      // Generate random number
+      const randomNumber = new RandomNumber();
+
+      // Encrypt random number with recipient's public key
+      const encrypted = viewingKeyPair.encrypt(randomNumber);
+
+      // Get x,y coordinates of ephemeral private key
+      const { pubKeyXCoordinate } = KeyPair.compressPublicKey(encrypted.ephemeralPublicKey);
+
+      // Compute stealth address
+      const stealthKeyPair = spendingKeyPair.mulPublicKey(randomNumber);
```

Comment:
We also talked about a helper method to do all of this logic, i.e. take the recipient, token address, and amount and return the stealth key pair, ciphertext, and pkx

Reply:
Checkout prepareSend helper method here 1b968b7f98e2a8dc3a68e91dbe806bde65b17568